src/test/java/picard/sam/CramCompatibilityTest.java

```diff
@@ -0,0 +1,319 @@
+import org.testng.annotations.Test;
+import picard.cmdline.CommandLineProgram;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
+public class CramCompatibilityTest {
+
+    public static final String CRAM_FILE = "testdata/picard/sam/test_cram_file.cram";
+    public static final String CRAM_FILE_2 = "testdata/picard/sam/test_cram_file_with_crai.cram";
+    public static final String CRAM_FILE_ONE_PAIR = "testdata/picard/sam/one_pair.cram";
+    public static final String CRAM_FILE_ONE_PAIR_MC = "testdata/picard/sam/MarkDuplicates/one_pair_mc.cram";
+
+    public static final String CRAM_FILE_QUERY_SORTED_UNMAPPED = "testdata/picard/sam/unmapped_queryname_sorted.cram";
+
+    public static final String REFERENCE_FILE = "testdata/picard/sam/test_cram_file.ref.fa";
+    public static final String FASTQ_FILE = "testdata/picard/sam/fastq2bam/fastq-sanger/5k-v1-Rhodobacter_LW1.sam.fastq";
+
+    public static final String CRAM_UNMAPPED = "testdata/picard/sam/SamFileConverterTest/unmapped.cram";
+
+    public static final String CRAM_UNMAPPED_PART_1 = "testdata/picard/sam/unmapped_part_1.cram";
+    public static final String CRAM_UNMAPPED_PART_2 = "testdata/picard/sam/unmapped_part_2.cram";
+
+    public static final String CRAM_SPLIT_UNMAPPED = "testdata/picard/sam/split_test_unmapped.cram";
+
+    public static final String MBA_ALIGNED_CRAM = "testdata/picard/sam/MergeBamAlignment/cliptest.aligned.cram";
+    public static final String MBA_UNMAPPED_CRAM = "testdata/picard/sam/MergeBamAlignment/cliptest.unmapped.cram";
+    public static final String MBA_REFERENCE = "testdata/picard/sam/MergeBamAlignment/cliptest.fasta";
+
+    @DataProvider(name = "programms")
+    public Object[][] getProgramWithParams() {
+        return new Object[][] {
+                { "picard.sam.AddOrReplaceReadGroups",
+                        "RGID=4 " +
+                        "RGLB=lib1 " +
+                        "RGPL=illumina " +
+                        "RGPU=unit1 " +
+                        "RGSM=20"
+                },
+                {"picard.sam.CleanSam", null},
+                {"picard.sam.DownsampleSam", null},
+                {"picard.sam.FixMateInformation", null},
+                {"picard.sam.markduplicates.MarkDuplicates", "M=m"},
+                {"picard.sam.MergeSamFiles", null},
+                {"picard.sam.PositionBasedDownsampleSam", "FRACTION=0.5"},
+                {"picard.sam.ReorderSam",
+                            "REFERENCE=" + REFERENCE_FILE + " " +
+                        "ALLOW_CONTIG_LENGTH_DISCORDANCE=true " +
+                        "ALLOW_INCOMPLETE_DICT_CONCORDANCE=true "
+                },
+                {"picard.sam.SortSam", "SORT_ORDER=coordinate"},
+        };
+    }
+
+    @Test(dataProvider = "programms")
+    public void writingMetricsShouldWriteCRAMWhenCRAMWithReferenceIsInput(String program, String exParameters)
```

Comment:
I think it would be clearer to not use the phrase metrics in these method names, as none of these are what I think of as metrics per se.